compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs

```diff
@@ -0,0 +1,404 @@
+        let last_supplied_span = self.gen_args.args[self.args_offset + self.provided - 1].span();
+        let span = first_supplied_span.with_hi(last_supplied_span.hi());
+
+        err.span_label(span, &format!(
+            "found {} {} argument{}",
+            self.provided,
+            self.kind,
+            pluralize!(self.provided),
+        ));
+
+        err.span_label(
+            span,
+            &format!(
+                "  and {} more {} expected",
+                missing_arg_count,
+                if missing_arg_count > 1 { "were" } else { "was" },
+            ),
+        );
+
+        self.suggest_adding_missing_args(err);
+    }
+
+    /// Prints diagnostics for when user has provided too many arguments (of given kind), e.g.:
+    ///
+    /// ```compile_fail
+    /// # use std::collections::HashMap;
+    /// type Map = HashMap<String, String, String, String>;
+    /// ```
+    fn handle_too_many_arguments(&self, err: &mut DiagnosticBuilder<'_>) {
+        assert!(self.provided > 0);
+
+        let unnecessary_args_from = {
+            // Index of the first unnecessary argument
+            let idx = self.args_offset + self.expected_max;
+
+            // When our unnecessary argument is second, third etc. on the generics list overall, we
+            // suggest to remove it together with its preceding comma, e.g.:
+            //
+            // ```
+            // type Map = HashMap<String, String, String, String>;
+            //                                          ^^^^^^^^
+            // ```
+            if idx > 0 {
+                self.gen_args.args[idx - 1].span().hi()
+            } else {
+                self.gen_args.args[idx].span().lo()
+            }
+        };
+
+        let unnecessary_args_to = {
+            // Index of the last unnecessary argument; may be equal to the previous index, but
+            // doesn't have to be
+            let idx = self.args_offset + self.provided - 1;
+
+            // Symmetrically to the case above - when our unnecessary argument is not the last one,
+            // we suggest to remove together with its following comma, e.g.:
+            //
+            // ```
+            // type Container = Box<'static, String>;
+            //                      ^^^^^^^^^
+            // ```
+            if let Some(arg) = self.gen_args.args.get(idx + 1) {
+                arg.span().lo()
+            } else {
+                self.gen_args.args[idx].span().hi()
+            }
+        };
+
+        let span = Span::new(unnecessary_args_from, unnecessary_args_to, self.span.ctxt());
+
+        let msg = if self.provided - self.expected_max > 1 {
+            format!("consider removing these {} arguments", self.kind)
+        } else {
+            format!("consider removing this {} argument", self.kind)
+        };
+
+        err.span_suggestion(span, &msg, String::new(), Applicability::MaybeIncorrect);
```

Comment:
I'm wondering if making this 
```suggestion
        err.span_suggestion_verbose(span, &msg, String::new(), Applicability::MaybeIncorrect);
```
will look better.

Reply:
Seems like `span_suggestion_verbose()` prints message with the suggestion already applied, yielding pearls such as:
```patch
@@ -19,7 +19,12 @@ LL | fn foo<const X: usize, const Y: usize>() -> usize {
    | ------------------------------------------------- defined here, with 2 const parameters: `X`, `Y`
 ...
 LL |     foo::<0, 0, 0>();
-   |     ^^^       --- help: remove this const argument
+   |     ^^^
+   |
+help: remove this const argument
+   |
+LL |     foo::<0, 0>();
+   |              --
```
or
```patch
@@ -25,7 +30,12 @@ LL | struct S;
    | --------- defined here, with 0 const parameters
 ...
 LL |     S::<0>;
-   |     ^   - help: remove this const argument
+   |     ^
+   |
+help: remove this const argument
+   |
+LL |     S::<>;
+   |        --
```
IMHO the inline (current) version with `s/consider replacing/replace` is way more legible.